oda_api/token.py

```diff
@@ -62,57 +68,154 @@ def decode_oda_token(token: str, secret_key=None, allow_invalid=False) -> dict:
+                                   "the currently available one")
+
+        current_decoded_token_roles = get_token_roles(current_decoded_token)
+        new_decoded_token_roles = get_token_roles(new_decoded_token)
+        new_roles_difference = set(new_decoded_token_roles) - set(current_decoded_token_roles)
+        current_roles_difference = set(current_decoded_token_roles) - set(new_decoded_token_roles)
+        if new_roles_difference != set() and current_roles_difference == set():
+            logger.warning("The new token has more roles than the current one:\n"
+                           f"roles current token: {current_decoded_token_roles}\n"
+                           f"roles new token: {new_decoded_token_roles}")
+        elif len(new_decoded_token_roles) < len(current_decoded_token_roles):
+            warning_msg = "The new token has less roles than the current one:\n" \
+                          f"roles current token: {current_decoded_token_roles}\n" \
+                          f"roles new token: {new_decoded_token_roles}\n"
+            if force_rewrite:
+                warning_msg += ", but it will be used.\n"
+                logger.warning(warning_msg)
+            else:
+                logger.warning(warning_msg)
+                raise RuntimeError("The roles of the new token are less than those of the current one")
+
+    if token_write_methods is not None:
+        if current_token is not None:
+            with open("old-oda-token_" + str(time.time()), 'w') as ft:
+                ft.write(current_token)
+
+        if isinstance(token_write_methods, TokenLocation):
+            token_write_methods = token_write_methods,
+
+        if discover_method is not None:
+            if discover_method == TokenLocation.ODA_ENV_VAR:
+                environ.pop('ODA_TOKEN', None)
+            elif discover_method == TokenLocation.FILE_CUR_DIR:
+                if path.exists(path.join(getcwd(), ".oda-token")):
+                    remove(path.join(getcwd(), ".oda-token"))
+            elif discover_method == TokenLocation.FILE_HOME:
+                if path.exists(path.join(environ["HOME"], ".oda-token")):
+                    remove(path.join(environ["HOME"], ".oda-token"))
+
+        for token_write_method in token_write_methods:
+            if token_write_method == TokenLocation.ODA_ENV_VAR:
+                environ['ODA_TOKEN'] = new_token
+            elif token_write_method == TokenLocation.FILE_CUR_DIR:
+                with open(path.join(getcwd(), ".oda-token"), 'w') as ft:
+                    ft.write(new_token)
+                chmod(path.join(getcwd(), ".oda-token"), 0o400)
+            elif token_write_method == TokenLocation.FILE_HOME:
+                with open(path.join(environ["HOME"], ".oda-token"), 'w') as ft:
+                    ft.write(new_token)
+                chmod(path.join(environ["HOME"], ".oda-token"), 0o400)
+
```

Comment:
here we can add a sanity check that `discover_token` returns the freshly written one. For consistency

Reply:
Sanity check added, and a warning is raised in case the discovered one is different from the refreshed one